### PR TITLE
feat: harden proactive delivery floor gating

### PR DESCRIPTION
**Pull request by CN-Zephyr**
## 改动概述 / Summary

本 PR 将通用宿主的主动投递收束为一个明确、不依赖插件实现的契约：

- **新鲜度**：每条 cue 可携带 `delivery_ttl_seconds`；宿主只建立一个跨 manager、callback 与 hot-swap 队列都不重置的单调时钟截止时间。
- **地板门禁**：只避免机械音频重叠。当前生成、前端播放、当前 server 音频，以及 client VAD 的 0.5 秒帧级防抖会阻止注入；不再把 8 秒 recent activity 或 6 秒 VAD grace 当成“安静窗口”。内容上的主动插话仍然允许。
- **一次消费**：同一 cue 在 direct injection 与 hot-swap priming 之间遵循 `waiting → claimed → committed`。只有一个路径能领取；首个不可逆原生媒体写入后进入 committed，重试只补齐配对文本，不重复媒体。
- **被动语义**：`ai_behavior=read` 仍然只入队，不主动触发生成；显式同 key 的 passive coalescing 保留。
- **隐私**：持久日志不记录原始 coalesce key 或 provider rejection 详情。

Refs #2491.

### 行为边界 / Behavioral Boundary

这个 PR 解决的是：

1. 已过期的 cue 不应在以后复活；
2. 用户或 AI 正在实际发声/播放时，不产生机械重叠；
3. 同一宿主队列项只能被一个投递路径消费一次；
4. 原生媒体已经写入 provider 后，失败/取消/迟到 rejection 不得留下无配对文本的孤立媒体，也不得在重试时重复媒体；
5. 告别静默期间已经进入持久队列的 cue 暂停截止时间，返回后再投递。

这个 PR **不**承诺长时间无打扰，也不把“最近说过话”视为仍在占用发言权。组长此前接受的主动插话语义保持不变。

### 改动前后 / Before and After

- Before: TTL 只在 manager 内生效，跨队列可能重启。After: producer TTL 变成一个绝对截止时间；缺失或无效 metadata 保持既有默认行为。
- Before: direct injection 与 hot-swap 各自读取队列快照。After: callback 与 mirror 共享 claim token，互斥领取并统一释放。
- Before: 媒体准备期间地板状态可能变化。After: 在 provider 注入前重新检查当前音频、播放、生成、session identity 与 hot-swap 状态，并在无后续生命周期事件时主动安排节流重试。
- Before: 首张原生图片成功、后续图片失败时，TTL 仍可能撤回文本。After: 首次不可逆媒体提交即固化 callback+mirror；重试跳过媒体，只完成文本。
- Before: 已释放到持久队列的 cue 若随后进入 goodbye silence，仍可能按旧 deadline 过期。After: 进入静默时同步暂停 callback 与配对 mirror 的 deadline。

### 潜在回归与覆盖 / Regression Risks

队列优先级/FIFO、显式 coalescing、TTL 跨队列、direct/hot-swap claim 互斥、取消与异常释放、媒体部分提交、迟到 rejection、goodbye silence、passive/read 语义，以及受影响的 `main_logic`、`memory`、`main_routers`、`app` 调用路径均有回归覆盖。

## 不拆分理由 / Why Not Split

本 PR 只有 9 个文件，但生产改动共同实现 queue admission、direct injection、native media commit 与 hot-swap consumption 的同一投递不变量；拆开会让中间状态缺少端到端互斥或提交语义。

## 范围 / Scope

PR 仍严格为 9 个文件。明确不包含：

- Live voice-passive 或任何插件专属行为；
- backend playback-completion callback；
- 插件可见 delivery/generation ID；
- compensation replay、brief-cue selection、bypass flag、battle-scoped queue control；
- Issue #2561 的 realtime WebSocket 断线诊断或修复。

provider 接受请求不等于真实播放完成；本 PR 不作这一保证。未知 metadata 继续安全透传并忽略。

## 测试 / Testing

- `uv run pytest -q tests/unit` — **8428 passed, 46 skipped**
- proactive/passive/hot-swap/passthrough 相关回归 — **285 passed**
- goodbye-silence 精确回归 — **7 passed**
- Ruff — passed
- `uv run python scripts/check_core_contracts.py` — passed
- `uv run python -m compileall -q main_logic app main_routers memory plugin/server/messaging/proactive_bridge.py` — passed
- `git diff --check` — passed

## Summary by CodeRabbit

* **新功能**
  * 支持在消息元数据中配置单条投递 TTL（仅接受正有限并封顶），基于单调时间生成截止点；到期后会被撤回并更新投递回执。
  * 语音“原生媒体不可逆提交”引入会话级提交绑定：同一会话可跳过重放，切换会话后可重新流式推送。
* **问题修复**
  * 热切换/交换（swap）期间对投递声明与被动回调的领取/释放/ACK 时序做了对齐，避免重复释放、错误过期清理及跨路径异常恢复。
* **日志安全性与稳定性**
  * 投递合并相关标识的日志展示更安全，避免输出完整合并键信息。

### CODE REVIEW

**Comment by CN-Zephyr** [comment]
Accepted the counter-separation suggestion. Commit 662ac355 gives unkeyed sentinel generation its own counter, keeps FIFO sequence numbers contiguous, and adds a regression test; focused test_proactive_delivery: 53 passed, Ruff/compileall/diff check passed. I also added the required 回归报告 / Regression Report section to the PR body based on the CI and review warning.

File: main_logic/core/proactive.py [hunk 1]
```diff
@@ -786,6 +945,37 @@ def _on_voice_inject_rejected(
...
+                    or self._voice_user_speaking(voice_sess)
+                ):
+                    self._release_delivery_claims(
+                        claimed_voice_snapshot, "direct"
+                    )
+                    logger.info(
+                        "[%s] trigger_agent_callbacks: final voice gate closed after media await",
+                        self.lanlan_name,
+                    )
                     return False
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in `7b240bc6`.

Once a callback has left the manager queue, a transient VAD/hot-swap closure no longer relies on an unrelated lifecycle event: both the initial inner voice gate and the post-media final gate now arm a paced retry using the manager min-gap. Every attempt rechecks expiry, coalescing, session/floor state, and the existing claim/lock prevents duplicate injection.

Validation: 153 focused proactive/passive/hot-swap tests passed; Ruff, compileall, and diff checks passed.

File: main_logic/core/lifecycle.py [hunk 2]
```diff
@@ -2054,13 +2059,10 @@ async def _perform_final_swap_sequence(self):
...
                     if not self._coalesce_entry_is_stale(e)
+                    and not callback_delivery_expired(e)
                 ]
-                final_prime_text += _render_pending_extra_replies_by_origin(
-                    _selected,
-                    lang=_lang,
-                    lanlan_name=self.lanlan_name,
-                    master_name=self.master_name,
-                )
+                _selected = self._claim_delivery_entries(_selected, "swap")
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/lifecycle.py]: Accepted and fixed in `ccabd172`.

Legacy plain-string extras now remain in the budget-selected swap batch while ownership claims are applied only to structured dict entries. They are rendered with the existing legacy fallback and removed only after a successful promote, preserving the established abort/restore semantics.

Validation: 154 focused proactive/passive/hot-swap tests passed; Ruff, compileall, and diff checks passed.

File: main_logic/core/proactive.py [hunk 3]
```diff
@@ -647,13 +795,17 @@ def _active_proactive_callbacks(callbacks: list) -> list:
                     self.state.phase is not ProactivePhase.IDLE
                     or voice_sess.is_active_response()
                     or self._is_voice_playing()
+                    or self._voice_user_speaking(voice_sess)
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:878]: Accepted; this is covered by `7b240bc6` (pushed before this review event was delivered).

Manager-released callbacks now arm a paced retry when the inner voice gate is closed, including VAD/recent-user activity and the post-media TOCTOU gate. Each attempt rechecks TTL/coalescing/session ownership, and the existing lock/claim prevents duplicate injection.

Validation is included in the current 154-test focused regression set.

**Comment by CN-Zephyr** [comment]
@codex审查

File: main_logic/core/proactive.py [hunk 4]
```diff
@@ -1856,6 +2089,11 @@ def enqueue_agent_callback(self, callback: dict) -> None:
             from config import (
                 AGENT_CALLBACK_QUEUE_MAX_ITEMS,
             )
+            ensure_callback_delivery_deadline(
+                callback,
+                default_ttl_s=None,
+            )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:2601]: Accepted and fixed in `5576b1f6`.

The passive hot-swap selector now runs the shared callback expiry sweep before candidate selection, matching the natural text-turn drain. Expired read callbacks are removed, their ack resolves `False`, and they cannot be rendered into a later swap prime.

Validation: 155 focused proactive/passive/hot-swap tests passed; Ruff, compileall, and diff checks passed.

**Comment by wehos** [comment]
@codex review

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 5]
```diff
@@ -570,6 +716,7 @@ def _active_proactive_callbacks(callbacks: list) -> list:
             "[%s] trigger_agent_callbacks enter: session=%s phase=%s pending=%d",
             self.lanlan_name, sess_type, self.state.phase.value, len(self.pending_agent_callbacks),
         )
+        self._purge_expired_agent_callbacks()
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:798]: Accepted and fixed in ce776622. The text/offline path now re-checks per-cue TTL after the final await boundary and before prompt_ephemeral. Expired entries are acked false and their paired extras/claim are cleared; a mixed batch renders only valid entries, while an all-expired batch releases inflight without prompting. Validation: 2 focused race tests, 151 focused regression tests, 760 broader proactive/topic tests, Ruff, compileall, and diff check.

File: main_logic/core/proactive.py [hunk 6]
```diff
@@ -786,6 +954,45 @@ def _on_voice_inject_rejected(
...
+                # Streaming media is the final await before provider injection.
+                # Re-check expiry and floor ownership here: VAD, playback or a
+                # hot swap may have changed while the media was in flight.
+                self._purge_expired_agent_callbacks()
+                voice_snapshot[:] = [
+                    cb
+                    for cb in voice_snapshot
+                    if not cb.get(DELIVERY_RETRACTED_KEY)
+                    and not callback_delivery_expired(cb)
+                ]
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in `aa751236`. TTL is now committed at the pre-stream boundary: once `stream_image()` has persisted the media conversation item, the post-upload path no longer drops the matching callback solely because its deadline crossed during the upload. This prevents stale media from being left for a later response without its explanatory text. Added `test_voice_mode_finishes_committed_media_when_ttl_crosses_during_upload` to lock the race.

Validation: 148 focused proactive/hot-swap tests passed; full `tests/unit` passed (`7136 passed, 43 skipped`); Ruff, compileall, and diff check passed.

File: main_logic/core/proactive.py [hunk 7]
```diff
@@ -647,24 +795,41 @@ def _active_proactive_callbacks(callbacks: list) -> list:
...
                     return False
 
                 _lang = normalize_language_code(self.user_language, format='short')
                 voice_snapshot = [
                     cb for cb in proactive_cbs
                     if not cb.get(DELIVERY_RETRACTED_KEY)
                 ]
+                voice_snapshot = self._claim_delivery_entries(
+                    voice_snapshot, "direct"
+                )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:920]: Accepted and fixed in `aa751236`. When the direct voice path finds that every callback is currently owned by the swap path, it now arms the existing paced retry before returning. The callback remains queued; after an aborted swap releases ownership, the retry can claim and deliver it without waiting for an unrelated lifecycle event. Added `test_voice_mode_swap_claim_denial_arms_retry`, including the release-then-deliver continuation.

Validation: 148 focused proactive/hot-swap tests passed; full `tests/unit` passed (`7136 passed, 43 skipped`); Ruff, compileall, and diff check passed.

File: tests/unit/test_proactive_sm_integration.py [hunk 8]
```diff
@@ -1476,6 +1563,123 @@ async def test_deliver_agent_callbacks_text_keeps_topic_hook_in_plain_text_sessi
...
+
+    assert delivered is True
+    assert len(sess.called_with) == 1
+    assert "valid body" in sess.called_with[0]
+    assert "expired body" not in sess.called_with[0]
+    assert expired_ack.result() is False
+    assert valid_ack.result() is True
+    assert mgr.pending_agent_callbacks == []
+    assert mgr.pending_extra_replies == []
+    assert expired_token not in mgr._proactive_delivery_claims
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply tests/unit/test_proactive_sm_integration.py]: 接受并已在 `aa751236` 修复。定向测试先证实原断言确实会留下成功条目的 `text` claim；因此不只是机械改测试：文本提交成功后现在复用 `_release_delivery_claims(...)` 清理有效条目，再把回归断言收紧为整个 claims 表为空。这样同时覆盖过期条目和成功条目的清理契约。

验证：相关 proactive/hot-swap 测试 148 passed；完整 `tests/unit` 为 `7136 passed, 43 skipped`；Ruff、compileall、diff check 均通过。

**Comment by CN-Zephyr** [comment]
@codex review

**Comment by CN-Zephyr** [comment]
Addressed the new CodeRabbit review (`pullrequestreview-4809760660`) in `2b7a05e7`: the swap-owned rejection-path test now also proves that the paired extra remains unchanged and ownership stays `{token: swap}` until the competing path releases it. This is test-only and does not change runtime behavior.

Validation: the exact regression passed; the complete `test_proactive_sm_integration.py` file passed (`75 passed`); Ruff, compileall, and diff check passed.

I also re-evaluated the earlier optional `_seq`/sentinel-counter nitpick from `pullrequestreview-4798468378` and am intentionally not changing it. `_QueuedCue.seq` is an ordering token whose contract is monotonicity, not contiguity; gaps from unique-key generation do not affect FIFO behavior. A second mutable counter would add production state solely for cosmetic log numbering, with no correctness or diagnosability benefit that justifies widening this PR.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/lifecycle.py [hunk 9]
```diff
@@ -2054,13 +2064,21 @@ async def _perform_final_swap_sequence(self):
                 _selected = [
                     e for e in _selected
                     if not self._coalesce_entry_is_stale(e)
+                    and not callback_delivery_expired(e)
                 ]
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/lifecycle.py:2233]: Accepted and fixed in `58668dea`. The swap path now purges stale/expired extras before `_select_callbacks_within_token_budget()`, so a superseded head entry cannot consume the selection window or indefinitely defer unrelated live extras. Added a regression that makes the selector admit only the first candidate and proves the stale entry is absent before that walk while the unrelated cue is delivered.

Validation: exact regressions 2 passed; `test_hot_swap_cancellation.py` 22 passed; related hot-swap/passive/text/context suite 147 passed; Ruff, compileall, and diff check passed. Full `tests/unit` reached 7,137 passed / 43 skipped with one unrelated launcher single-instance failure caused by the currently running legacy N.E.K.O lock.

File: main_logic/core/lifecycle.py [hunk 10]
```diff
@@ -2316,6 +2364,28 @@ async def _perform_final_swap_sequence(self):
...
+                    cb
+                    for cb in self.pending_agent_callbacks
+                    if cb.get(DELIVERY_CLAIM_TOKEN_KEY) not in delivered_tokens
+                ]
+                for callback in delivered_callbacks:
+                    resolve_callback_delivery_ack(callback, True)
+                self._release_delivery_claims(_swap_claimed_extras, "swap")
+                _swap_claim_committed = True
+                _removed_extras = []
+                _removed_cb_backed_ids = set()
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/lifecycle.py]: Accepted and fixed in `58668dea`. Delivery ack/removal, claim release, and the committed marker now occur only after the final awaited preparation-state reset completes. Until then `_removed_extras` and the paired callback remain restorable, so cancellation in that last cleanup window requeues the cue, leaves its ack unresolved, and releases swap ownership in `finally`.

The new regression cancels specifically inside the final reset boundary and verifies queue restoration, callback preservation, no successful ack, and an empty claim ledger. Validation: exact regressions 2 passed; `test_hot_swap_cancellation.py` 22 passed; related suite 147 passed; Ruff, compileall, and diff check passed. Full `tests/unit` reached 7,137 passed / 43 skipped with one unrelated active-runtime single-instance test failure.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/lifecycle.py [hunk 11]
```diff
@@ -2054,13 +2074,21 @@ async def _perform_final_swap_sequence(self):
...
-                    lang=_lang,
-                    lanlan_name=self.lanlan_name,
-                    master_name=self.master_name,
+                # Legacy plain-string extras are still supported by both the
+                # budget selector and renderer, but they have no delivery
+                # identity to claim. Preserve them in the selected batch while
+                # applying direct-vs-swap ownership only to structured entries.
+                _claimed_selected = self._claim_delivery_entries(
+                    [e for e in _selected if isinstance(e, dict)], "swap"
                 )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/lifecycle.py:2240]: Accepted and fixed in `68b627cb`. Hot-swap budget candidates now exclude structured extras owned by another delivery path before the budget walk; claiming still happens only after selection, so unowned deferred entries are not pre-claimed. The direct-vs-swap regression forces a one-item budget window and proves the direct-owned head entry is absent while the unrelated cue is delivered and direct ownership remains intact. Validation: exact test 1 passed; hot-swap file 22 passed; related suite 147 passed; Ruff, compileall, and diff check passed.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 12]
```diff
@@ -37,6 +45,146 @@
...
+        now = time.monotonic()
+        claims = getattr(self, "_proactive_delivery_claims", {})
+        expired_tokens: set[object] = set()
+        kept_callbacks: list = []
+        expired_count = 0
+        for callback in getattr(self, "pending_agent_callbacks", []):
+            if isinstance(callback, dict) and callback_delivery_expired(
+                callback, now=now
+            ):
+                expire_callback_delivery(callback)
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:144]: Accepted and fixed in `28d46891`. The shared queue expiry purge now defers callbacks and paired extras while their delivery token is actively owned, so a concurrent retry cannot resolve the ack `False` or remove a cue already primed by swap. After the owner commits or releases the claim, the existing owner/expiry paths resume normally. The regression proves the expired callback+mirror and pending ack survive a `swap` claim, then expire and ack `False` only after release. Validation: exact regression 1 passed; related proactive/hot-swap/passive suite 162 passed; Ruff, compileall, and diff check passed.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 13]
```diff
@@ -570,6 +750,7 @@ def _active_proactive_callbacks(callbacks: list) -> list:
             "[%s] trigger_agent_callbacks enter: session=%s phase=%s pending=%d",
             self.lanlan_name, sess_type, self.state.phase.value, len(self.pending_agent_callbacks),
         )
+        self._purge_expired_agent_callbacks()
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:798]: Accepted and fixed in e1537cf5. Callbacks enqueued while goodbye-silent now drop their delivery deadline before entering the persistent queue, so a deferred retry cannot expire/ack-false either a manager-parked cue or its paired hot-swap mirror. Normal (non-goodbye) TTL handling is unchanged. Added regressions for explicit short TTL and an already-expired manager callback; related proactive suites pass (130 tests), plus Ruff, compileall, and diff check.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 14]
```diff
@@ -786,6 +997,46 @@ def _on_voice_inject_rejected(
...
+                        claimed_voice_snapshot, "direct"
+                    )
+                    # Media streaming is the last await before provider send.
+                    # If user activity or a hot swap closes the gate during that
+                    # await, no response was created and therefore no
+                    # response.done / playback-end callback is guaranteed to
+                    # retry this manager-released cue.
+                    self._schedule_proactive_retry(
+                        self.proactive_manager.min_gap_s
+                    )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in 48943ec5. After a successful media stream, only media-bearing callbacks and their paired extras now clear the delivery deadline, so a floor/hot-swap deferral cannot expire the matching text on retry; non-media callbacks keep their original TTL. Added a regression covering TTL crossing during upload, the post-upload voice gate closing, no premature inject, and successful retry. Validation: 130 related tests passed; Ruff, compileall, and diff check passed.

**Comment by CN-Zephyr** [comment]
@codex review

**Comment by CN-Zephyr** [comment]
Accepted and fixed in `c1e52193`. The direct delivery claim is now released on all four pre-media early exits identified in `pullrequestreview-4814446745`: stale snapshot before SID rotation, SID rotation failure, activity starting during rotation, and staleness after rotation. This keeps the callback retryable/expirable and allows hot-swap to claim it after the direct path defers.

Added regressions for both pre-rotation and during-rotation retraction, and strengthened the existing rotation-failure/activity tests to assert the claim map is empty. I intentionally kept the change to the four exits instead of introducing a context-manager refactor, which would broaden this conflict-resolution follow-up beyond the reported defect.

Validation: 4 exact regressions passed; the related proactive/hot-swap suite passed (`230 passed`); Ruff, py_compile, and diff check passed.

File: main_logic/core/proactive.py [hunk 15]
```diff
@@ -921,9 +1141,64 @@ def _on_voice_media_rejected(error_msg: str) -> None:
...
+                    for cb in voice_snapshot
+                    if cb.get("media_images")
+                    and cb.get("_callback_delivery_id")
+                }
+                for cb in voice_snapshot:
+                    if cb.get("media_images"):
+                        cb.pop(DELIVERY_DEADLINE_KEY, None)
+                for extra in voice_extra_snapshot:
+                    if extra.get("_callback_delivery_id") in committed_media_ids:
+                        extra.pop(DELIVERY_DEADLINE_KEY, None)
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Thanks; I’m not changing this behavior. The deadline removal is deliberate and scoped only to media-bearing callbacks (and their paired extras) after `_stream_cb_media()` succeeds. At that point the realtime provider already contains irreversible media. Letting the original TTL expire would drop the matching callback text and leave unexplained stale media, which is exactly the previously fixed Codex P2 in discussion_r3676457712. Non-media callbacks retain their TTL. The durable retry is the atomic media+text delivery contract, not a general freshness bypass; the regression covers TTL crossing during upload, the post-upload gate closing, no early inject, and eventual matching delivery.

File: main_logic/core/proactive.py [hunk 16]
```diff
@@ -862,6 +1073,9 @@ def _on_voice_media_rejected(error_msg: str) -> None:
                         )
                         # No response.create fired, so no response lifecycle
                         # hook can re-drive this retained callback.
+                        self._release_delivery_claims(
+                            claimed_voice_snapshot, "direct"
+                        )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in `35289c54`. Direct voice claims are now released before cancellation/exception propagation from SID rotation, media streaming, pre-inject preparation, and the injection await. Added regressions for cancellation at SID rotation, media streaming, and injection; each verifies propagation, callback retention, committed-marker cleanup, and an empty claim ledger. Validation: 4 exact regressions passed; 222 related proactive/realtime/hot-swap/passive tests passed; Ruff, py_compile, and diff checks passed.

File: main_logic/core/proactive.py [hunk 17]
```diff
@@ -878,6 +1092,9 @@ def _on_voice_media_rejected(error_msg: str) -> None:
                         "[%s] trigger_agent_callbacks: activity started during SID rotation; deferring callback delivery",
                         self.lanlan_name,
                     )
+                    self._release_delivery_claims(
+                        claimed_voice_snapshot, "direct"
+                    )
                     return False
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in `35289c54`. When user/AI activity starts during SID rotation, the direct claim is released and the existing paced retry is now armed before returning, so the manager-released callback no longer depends on an unrelated lifecycle event. The existing rotation-activity regression now asserts the retry schedule as well as callback retention and claim cleanup. Validation: 4 exact regressions passed; 222 related tests passed; Ruff, py_compile, and diff checks passed.

**Comment by CN-Zephyr** [comment]
Addressed all three outside-diff findings from `pullrequestreview-4814551768` in `599000b2`:

1. Accepted: after the final media/rotation await, callbacks that still retain a deadline (non-media cues) are re-checked with the monotonic clock before inject. Media-bearing callbacks keep the previously established durable media+text contract after successful provider persistence.
2. Accepted: legacy plain-string extras are preserved; all affected voice/text snapshot, retraction, rejection-dedup, and success-cleanup scans now call `.get()` only on structured dict entries.
3. Accepted: inflight expiry releases only the current delivery path’s claim. I also completed the existing ownership contract by actually claiming text callbacks before removing them from the host queue (the text release path already existed), so a swap/direct-owned cue stays queued and cannot be expired or delivered by text concurrently.

Regression coverage includes non-media TTL crossing during SID rotation, committed-media durability, legacy-string coexistence, foreign-owner expiry protection, and text-vs-swap claim exclusion. Validation: 7 exact regressions passed; 226 focused proactive/realtime/hot-swap/passive tests passed; 911 expanded proactive/topic tests passed; Ruff, py_compile, Core contracts, and diff checks passed. PR scope remains the same 9 files.

**Comment by CN-Zephyr** [comment]
@codex审查

**Comment by CN-Zephyr** [comment]
进过重新审核后发现修改范围漂移，正在重新修改

File: main_logic/core/proactive.py [hunk 18]
```diff
@@ -921,9 +1164,75 @@ def _on_voice_media_rejected(error_msg: str) -> None:
...
+                for cb in voice_snapshot:
+                    if cb.get("media_images"):
+                        cb.pop(DELIVERY_DEADLINE_KEY, None)
+                for extra in voice_extra_snapshot:
+                    if extra.get("_callback_delivery_id") in committed_media_ids:
+                        extra.pop(DELIVERY_DEADLINE_KEY, None)
+                expired_count = self._expire_inflight_callback_snapshots(
+                    voice_snapshot,
+                    path="direct",
+                )
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:1228]: Accepted and refactored in `6426f304`. The first successful native media write now commits the callback and its paired mirror, removes their deadlines, and makes later retries skip media while completing only the matching text. This covers a second-image failure or cancellation without introducing a per-image lifecycle ledger. The unified attempt cleanup also releases the direct claim on every early exit. Regression coverage includes first-image success followed by failure/cancellation across TTL; the 282-test related suite and full 8428-test unit suite pass.

File: main_logic/core/proactive.py [hunk 19]
```diff
@@ -587,6 +776,7 @@ def _active_proactive_callbacks(callbacks: list) -> list:
             "[%s] trigger_agent_callbacks enter: session=%s phase=%s pending=%d",
             self.lanlan_name, sess_type, self.state.phase.value, len(self.pending_agent_callbacks),
         )
+        self._purge_expired_agent_callbacks()
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:798]: Accepted and fixed in `96fbc15d`. Entering goodbye silence now suspends deadlines for callbacks that had already reached `pending_agent_callbacks` and for their paired `pending_extra_replies` mirrors before the manager-held callbacks are parked. This preserves the existing cat-mode deferral contract without changing normal TTL behavior. Added an exact regression with an already-expired pending callback; it remains queued, unacknowledged, and deadline-free while silent. Validation: 7 goodbye tests and the 282-test related suite pass; Ruff and core contract checks pass.

**Comment by CN-Zephyr** [comment]
Accepted the latest CodeRabbit outside-diff finding and fixed it in `4c00f2ef`.

The flood-guard mirror filter now type-checks legacy plain-string entries before reading `_callback_delivery_id`. It preserves those legacy entries while removing only the structured mirror paired with a dropped callback, so the outer best-effort guard no longer swallows `AttributeError` and skips cleanup.

Added a regression covering a mixed legacy-string/structured-extra queue overflowing its callback limit. Validation: `test_proactive_delivery.py` 55 passed; related proactive/passive/hot-swap/passthrough suite 283 passed; Ruff and core contract checks passed.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/lifecycle.py [hunk 20]
```diff
@@ -1951,6 +1953,11 @@ def _select_passive_callbacks_for_swap_prime(self, extras_selected: list = None)
         gates that this path must not bypass.
         """
         try:
+            # TTL is a delivery-wide freshness contract.  Passive callbacks
+            # must cross the same expiry boundary here as they do in the
+            # natural text-turn drain, otherwise an old read snapshot can
+            # resurface when the next voice hot swap primes its new session.
+            self._purge_expired_agent_callbacks()
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/lifecycle.py:1992]: Accepted and fixed in `04f07ec9`. Passive/read callbacks selected for hot-swap priming are now atomically claimed by the existing `swap` owner before rendering and remain claim-protected through the prime/promote window. TTL purge and competing delivery paths therefore cannot ack/drop a callback after its text is already primed. Every success, abort, render failure, cancellation, and restore path releases the claim through the existing cleanup. Added an exact regression that expires the TTL and invokes purge inside `prime_context()`: the callback stays pending and unacked while claimed, is acked true only after promote, and leaves an empty claim ledger. Validation: 35 swap tests and the 285-test related suite pass; Ruff and core contracts pass.

**Comment by CN-Zephyr** [comment]
Partially accepted the latest CodeRabbit outside-diff finding and fixed the valid integrity issue in `9a479b92`.

Accepted: swap's hand-written expiry filters now preserve every actively claimed structured mirror, matching `_purge_expired_agent_callbacks()`. A direct-owned callback/mirror pair therefore remains intact while direct delivery owns the shared token, and the swap budget still excludes the `direct` owner.

Not adopted: making an already-expired cue retryable after an unsuccessful claim is released. That would revive stale producer data and violate this PR's TTL contract. The regression pins the intended boundary: while claimed, the expired callback/mirror pair is retained and unacked; after the failed claim is released, the next normal purge removes both halves together and resolves ack false.

Validation: exact regression passed; 35 swap tests and the 285-test related suite passed; Ruff and core contracts passed.

**Comment by CN-Zephyr** [comment]
Accepted the latest CodeRabbit outside-diff finding and fixed it in `1e6af866`.

Passive callbacks are now removed at promote without completing their delivery ACK. ACK true is deferred to the final cleanup commit boundary. Post-promote cancellation or a dead promoted session restores the callback with its ACK still pending; a surviving promoted session that cannot restore commits the ACK to avoid leaving the producer hanging.

Regression coverage pins successful final commit, post-promote cancellation restore, and ws-invalid/dead-session restore. Validation: 4 exact ACK-boundary cases passed; 35 swap tests and the 285-test related suite passed; Ruff and core contracts passed.

**Comment by CN-Zephyr** [comment]
Accepted the latest CodeRabbit outside-diff finding and fixed it in `39585faa`.

The normal extras commit path is now a reusable transaction-finalization helper. When promote has already removed extras but a later pre-commit step raises while the promoted session remains alive, the exception path now performs the same finalization: it removes callbacks by the shared delivery token, resolves their ACKs true, and releases the swap claims. Dead-session and cancellation exits still restore the uncommitted delivery instead, so retry and no-double-delivery boundaries are unchanged.

The existing post-promote/live-session regression now includes a paired callback and verifies that the extra is not restored, the callback is consumed, ACK resolves true, and no claim remains. Validation: exact regression passed; 35 swap/passive tests and the 285-test related suite passed; Ruff, core contracts, compileall, and diff checks passed.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 21]
```diff
@@ -697,9 +956,30 @@ def _active_proactive_callbacks(callbacks: list) -> list:
...
+                            if extra.get("_callback_delivery_id")
+                            == delivery_id
+                        )
+                    return entries
+
+                def _commit_native_media(callback: dict) -> None:
+                    """Make paired text durable after the first provider write."""
+                    for entry in _entries_for_voice_callback(callback):
+                        entry[VOICE_DELIVERY_COMMITTED_KEY] = True
+                        entry.pop(DELIVERY_DEADLINE_KEY, None)
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py]: Accepted and fixed in `da107d69`.

The durable media marker now records the realtime-session token that received the native image. A retry on that same session still skips replay, preserving the partial-commit contract; a retry after hot-swap promotion sees a different token and re-streams the image before injecting callback text. Clearing the committed state removes both markers.

A regression switches sessions inside the media await and verifies the first attempt defers, the promoted session receives the image again, and only then receives/prunes the callback. Validation: 4 focused media-boundary tests and the 286-test related suite passed; Ruff, core contracts, compileall, and diff checks passed.

**Comment by CN-Zephyr** [comment]
Accepted the latest CodeRabbit nitpick and fixed it in `0f42ed93`.

The new session-swap media regression now imports and uses `VOICE_DELIVERY_COMMITTED_KEY` instead of duplicating its string value. This is test-only and does not change runtime behavior.

Validation: the full 102-test proactive state-machine integration file passed; Ruff and diff checks passed.

**Comment by CN-Zephyr** [comment]
@codex review

File: main_logic/core/proactive.py [hunk 22]
```diff
@@ -746,13 +1030,16 @@ def _on_voice_inject_rejected(
                         if not cb.get(DELIVERY_RETRACTED_KEY)
                     ]
                     _state["rejected"] = True
+                    self._release_delivery_claims(_snapshot, "direct")
```
[the remark this answers is not in the document]
Comment (CN-Zephyr) [reply main_logic/core/proactive.py:1033]: Accepted and fixed in `3cdc2e5f`.

Media-event rejection is now distinct from `response.create` rejection. The callback keeps its durable transaction marker (so TTL/coalescing cannot invalidate an in-flight partial commit), but the rejected image clears only its committed-session token, forcing the queued retry to re-stream on that same session. A rejection observed before `stream_image()` returns also suppresses the subsequent native-commit mark. Normal response-create rejection still preserves successfully persisted media.

Regression coverage pins both synchronous pre-inject rejection and late asynchronous rejection followed by a same-session image replay. Validation: 2 focused rejection tests, the full 102-test state-machine file, and the 286-test related suite passed; Ruff, core contracts, compileall, and diff checks passed.
